File: backend/utils/render_env.py
"""
Render Environment Variable Utilities

Render provides service URLs as 'host' only (e.g., 'my-service.onrender.com')
This utility helps construct full URLs with https:// prefix.
"""
import os
from typing import Optional, Dict
from urllib.parse import urlparse, urlunparse


def get_full_url_from_host(host: Optional[str], scheme: str = "https") -> Optional[str]:
    """
    Convert a host to a full URL with scheme.
    
    Args:
        host: The hostname (e.g., 'my-service.onrender.com')
        scheme: The URL scheme to use (default: 'https')
    
    Returns:
        Full URL with scheme, or None if host is None
    """
    if not host:
        return None
    
    # Check if host already has a scheme
    if "://" in host:
        return host
    
    # Construct full URL
    return f"{scheme}://{host}"


def get_render_service_url(env_var: str, scheme: str = "https") -> Optional[str]:
    """
    Get a full URL from a Render environment variable that contains only a host.
    
    Args:
        env_var: The environment variable name
        scheme: The URL scheme to use (default: 'https')
    
    Returns:
        Full URL with scheme, or None if env var is not set
    """
    host = os.getenv(env_var)
    return get_full_url_from_host(host, scheme)


def get_render_urls() -> Dict[str, Optional[str]]:
    """
    Get all Render service URLs as full URLs.
    
    Returns:
        Dictionary of service names to their full URLs
    """
    return {
        "frontend": get_render_service_url("FRONTEND_URL"),
        "api": get_render_service_url("NEXT_PUBLIC_API_URL"),
        "nextauth": get_render_service_url("NEXTAUTH_URL"),
        "dashboard": get_render_service_url("DASHBOARD_URL"),
    }


def construct_service_endpoint(base_host: Optional[str], path: str = "", scheme: str = "https") -> Optional[str]:
    """
    Construct a full service endpoint URL.
    
    Args:
        base_host: The base hostname
        path: The path to append (e.g., '/api/health')
        scheme: The URL scheme to use (default: 'https')
    
    Returns:
        Full endpoint URL, or None if base_host is None
    """
    base_url = get_full_url_from_host(base_host, scheme)
    if not base_url:
        return None
    
    # Ensure path starts with /
    if path and not path.startswith("/"):
        path = f"/{path}"
    
    return f"{base_url}{path}"


# Convenience functions for specific services
def get_frontend_url(path: str = "") -> Optional[str]:
    """Get the frontend service URL with optional path."""
    host = os.getenv("FRONTEND_URL")
    return construct_service_endpoint(host, path)


def get_api_url(path: str = "") -> Optional[str]:
    """Get the API service URL with optional path."""
    host = os.getenv("NEXT_PUBLIC_API_URL")
    return construct_service_endpoint(host, path)


def get_dashboard_url(path: str = "") -> Optional[str]:
    """Get the dashboard service URL with optional path."""
    host = os.getenv("DASHBOARD_URL")
    return construct_service_endpoint(host, path)


# Example usage in your application:
if __name__ == "__main__":
    # Get full URLs
    frontend_url = get_frontend_url()  # https://me-core-dashboard.onrender.com
    api_health = get_api_url("/api/health")  # https://me-core-backend.onrender.com/api/health
    
    # Get all URLs
    urls = get_render_urls()
    print(f"Frontend URL: {urls['frontend']}")
    print(f"API URL: {urls['api']}")

File: dashboard/src/utils/render-env.ts
/**
 * Render environment utilities for handling host-only environment variables.
 * 
 * Render provides host-only values (e.g., 'example.onrender.com') for service URLs.
 * This module provides utilities to construct full URLs from these hosts.
 */

/**
 * Convert a host-only value to a full URL.
 * 
 * @param host - The host value (e.g., 'example.onrender.com')
 * @param defaultScheme - The scheme to use (default: 'https')
 * @returns Full URL (e.g., 'https://example.onrender.com') or null if host is null/empty
 */
export function getFullUrlFromHost(host: string | null | undefined, defaultScheme: string = 'https'): string | null {
  if (!host) {
    return null;
  }
  
  // If it already has a scheme, return as-is
  if (host.startsWith('http://') || host.startsWith('https://')) {
    return host;
  }
  
  // Add the scheme
  return `${defaultScheme}://${host}`;
}

/**
 * Get a full URL from a Render host environment variable.
 * 
 * @param envVarName - Name of the environment variable containing the host
 * @param defaultScheme - The scheme to use (default: 'https')
 * @returns Full URL or null if environment variable is not set
 */
export function getRenderServiceUrl(envVarName: string, defaultScheme: string = 'https'): string | null {
  const host = process.env[envVarName];
  return getFullUrlFromHost(host, defaultScheme);
}

/**
 * Get all Render service URLs from environment variables.
 * 
 * @returns Object containing service names mapped to full URLs
 */
export function getRenderUrls() {
  return {
    backend: getRenderServiceUrl('NEXT_PUBLIC_RENDER_BACKEND_HOST'),
    markerEngine: getRenderServiceUrl('NEXT_PUBLIC_RENDER_MARKER_ENGINE_HOST'),
    dashboard: getRenderServiceUrl('NEXT_PUBLIC_RENDER_DASHBOARD_HOST'),
    repairService: getRenderServiceUrl('NEXT_PUBLIC_RENDER_REPAIR_SERVICE_HOST'),
  };
}

/**
 * Construct a full service endpoint URL with path.
 * 
 * @param serviceHostEnv - Environment variable name containing the host
 * @param path - Path to append to the URL (should start with /)
 * @param defaultScheme - The scheme to use (default: 'https')
 * @returns Full URL with path or null if host is not set
 */
export function constructServiceEndpoint(
  serviceHostEnv: string, 
  path: string = '', 
  defaultScheme: string = 'https'
): string | null {
  const baseUrl = getRenderServiceUrl(serviceHostEnv, defaultScheme);
  if (!baseUrl) {
    return null;
  }
  
  // Ensure path starts with /
  if (path && !path.startsWith('/')) {
    path = `/${path}`;
  }
  
  return `${baseUrl}${path}`;
}

// Convenience functions for specific services
export function getBackendUrl(path: string = ''): string | null {
  return constructServiceEndpoint('NEXT_PUBLIC_RENDER_BACKEND_HOST', path);
}

export function getMarkerEngineUrl(path: string = ''): string | null {
  return constructServiceEndpoint('NEXT_PUBLIC_RENDER_MARKER_ENGINE_HOST', path);
}

export function getDashboardUrl(path: string = ''): string | null {
  return constructServiceEndpoint('NEXT_PUBLIC_RENDER_DASHBOARD_HOST', path);
}

export function getRepairServiceUrl(path: string = ''): string | null {
  return constructServiceEndpoint('NEXT_PUBLIC_RENDER_REPAIR_SERVICE_HOST', path);
}

// For backward compatibility with existing code
export function getServiceUrl(serviceName: 'backend' | 'markerEngine' | 'dashboard' | 'repairService'): string | null {
  const urlGetters = {
    backend: getBackendUrl,
    markerEngine: getMarkerEngineUrl,
    dashboard: getDashboardUrl,
    repairService: getRepairServiceUrl,
  };
  
  const getter = urlGetters[serviceName];
  return getter ? getter() : null;
}

// Type definitions for service URLs
export interface RenderServiceUrls {
  backend: string | null;
  markerEngine: string | null;
  dashboard: string | null;
  repairService: string | null;
}

// Helper to get non-null URLs (throws if any required service is missing)
export function getRequiredRenderUrls(): Required<RenderServiceUrls> {
  const urls = getRenderUrls();
  
  for (const [service, url] of Object.entries(urls)) {
    if (!url) {
      throw new Error(`Missing required Render service URL for ${service}. Please set NEXT_PUBLIC_RENDER_${service.toUpperCase()}_HOST environment variable.`);
    }
  }
  
  return urls as Required<RenderServiceUrls>;
}

// Helper to validate all required URLs are present
export function validateRenderUrls(): boolean {
  const urls = getRenderUrls();
  return Object.values(urls).every(url => url !== null);
}

// Export a default instance with all URLs
export const renderUrls = getRenderUrls();